src/transformers/models/llama/modeling_llama.py

```diff
@@ -699,28 +640,32 @@ def forward(
         value_states = value_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
 
         kv_seq_len = key_states.shape[-2]
+        past_seen_tokens = 0
+        past_key_value = getattr(self, "past_key_value", past_key_value)
         if past_key_value is not None:
-            kv_seq_len += past_key_value.get_usable_length(kv_seq_len, self.layer_idx)
+            past_seen_tokens = past_key_value.get_usable_length(kv_seq_len, self.layer_idx)  # add what was seen
+            kv_seq_len += past_seen_tokens
         cos, sin = self.rotary_emb(value_states, seq_len=kv_seq_len)
 
+        new_cache_positions = torch.arange(past_seen_tokens, past_seen_tokens + q_len, device=key_states.device)
```

Comment:
I did not benchmarked that yet, but it seems this might create a considerable overhead for FA2 models as for each attention module on each layer you need to re-create `new_cache_positions` and this might become quite an overhead for large sequence length. I will do some benchmarks and get back

Reply:
Yes, but hear me out. It's a fairly small tensor. Either you create it to slice, or you don't but you slice with a list, which will induce device copy. 
So we have 1 solution:
- pass the `cache positions` as inputs. 
- pass the `postiions_ids` as inputs. 
This means we have to keep track of the real positions, and the cache positions. 
Also my benchmark give no overhead for this